# Extending Neural Embeddings: From Words to Sentences, Documents, and Code

## 1. Is it Possible?
Yes, it is entirely possible and highly effective to extend neural embeddings beyond single words or tokens to represent entire sentences, documents, and programming functions. 

While early models like Word2Vec and GloVe focused on word-level embeddings, modern NLP has shifted towards models that can encode larger, unbounded sequences of text into fixed-length, dense vectors. These extended embeddings power search engines, recommendation systems, semantic code search, and Retrieval-Augmented Generation (RAG) pipelines.

## 2. The Model & 3. Architecture

To achieve this, we rely on architectures designed for sequence-level embeddings. 
- **For Sentences/Documents:** Models like **Sentence-BERT (SBERT)** are typically used. They produce a single continuous vector for an entire sentence or document.
- **For Programming Functions:** Models like **CodeBERT** are trained on massive corpora of programming languages paired with natural language, allowing them to embed both into the same vector space.

### Siamese Network Architecture
SBERT uses a **Siamese Network Classification Architecture**. 
1. **Twin Networks:** Two independent BERT models (with tied weights) process two inputs (e.g., "query" and "document").
2. **Pooling Layer:** The network computes a pooling operation, typically **Mean Pooling** (averaging the vectors of all output tokens), to derive a single fixed-size sentence embedding.
3. **Objective:** During training, the cosine similarity between the two embeddings is computed.

## 4. Code Implementation

In [3]:
def add_vectors(v1, v2):
    return [x + y for x, y in zip(v1, v2)]

In [4]:
# Install dependencies if needed
# !pip install sentence-transformers

from sentence_transformers import SentenceTransformer, util

# Load a pre-trained model capable of both text and code representations
model = SentenceTransformer('all-MiniLM-L6-v2')

# 1. Sentence/Document examples
doc1 = "Neural embeddings convert text to vectors."
doc2 = "Word vectors represent the semantic meaning of tokens."

# 2. Programming Function (Python code snippet example)
code_snippet = add_vectors

# Encode all sequences into dense vectors
embeddings = model.encode([doc1, doc2, code_snippet])

print(f"Generated Embeddings Shape: {embeddings.shape}")


Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


Generated Embeddings Shape: (3, 384)


In [5]:
# Compute Cosine Similarity between document 1 and document 2
cosine_sim = util.cos_sim(embeddings[0], embeddings[1])
print(f"Similarity between Doc 1 and Doc 2: {cosine_sim.item():.4f}")

# Compare the code snippet to standard text
code_sim = util.cos_sim(embeddings[0], embeddings[2])
print(f"Similarity between Doc 1 and Code Snippet: {code_sim.item():.4f}")


Similarity between Doc 1 and Doc 2: 0.5428
Similarity between Doc 1 and Code Snippet: 0.3005


## 5. Evaluation

How do we know if these embeddings are good?
1. **Intrinsic Evaluation:** We measure the cosine similarity between embeddings of pairs of sentences or code snippets that humans have manually scored for similarity (e.g., checking it against STS benchmarks).
2. **Extrinsic Evaluation:** We use the embeddings in a downstream task, such as a **Semantic Code Search Engine**. If the embeddings successfully retrieve the correct Python function for a user's natural language query, the model is effective.

## 6. Memory and Storage
Storing dense vectors for scale requires significant memory footprint.
- Our model produces a float32 vector of size 384.
- 1 million embeddings = 1,000,000 * 384 * 4 bytes ≈ **1.5 GB in RAM**.
To manage this, we use **Vector Databases** (Pinecone, Qdrant) or indexing libraries like **FAISS**, which apply quantization algorithms to compress vectors and perform rapid Approximate Nearest Neighbor (ANN) searches.

## 7. Guardrails
When applying these models, specific guardrails are critical:
- **Context Length Limits:** Models have token limits (e.g., 512 tokens). If a function is too long, the model silently truncates it. **Guardrail:** Implement chunking strategies before embedding long files.
- **Bias:** Models trained on code repositories may inherit developer biases or toxic comments from codebases. **Guardrail:** Continuously benchmark models and apply moderation APIs to the inputs/outputs.
- **Domain Shifts:** A standard text model won't represent C++ code well. **Guardrail:** Ensure the model matches the domain; use CodeBERT specifically for code retrieval tasks.